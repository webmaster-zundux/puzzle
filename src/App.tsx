import "@fontsource-variable/open-sans";
import type { FC } from "react";
import { memo } from "react";
import "./styles.css";

import { PuzzleDataProvider } from "./components/PuzzleDataProvider";
import { PuzzlePage } from "./pages/PuzzlePage";

export const App: FC = memo(function App() {
  return (
    <PuzzleDataProvider>
      <PuzzlePage />
    </PuzzleDataProvider>
  );
});
