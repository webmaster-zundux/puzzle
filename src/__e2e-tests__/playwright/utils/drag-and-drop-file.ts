import type { Locator, Page } from "@playwright/test";
import { readFileSync } from "fs";

export const dragAndDropFileByLocator = async (
  page: Page,
  locator: Locator,
  filePath: string,
  fileName: string,
  fileType: string,
) => {
  const buffer = readFileSync(filePath).toString("base64");

  const dataTransfer = await page.evaluateHandle(
    async ({ bufferData, localFileName, localFileType }) => {
      const dt = new DataTransfer();
      const blobData = await fetch(bufferData).then((res) => res.blob());
      const file = new File([blobData], localFileName, { type: localFileType });
      dt.items.add(file);

      return dt;
    },
    {
      bufferData: `data:application/octet-stream;base64,${buffer}`,
      localFileName: fileName,
      localFileType: fileType,
    },
  );

  await locator.dispatchEvent("drop", { dataTransfer });
};
