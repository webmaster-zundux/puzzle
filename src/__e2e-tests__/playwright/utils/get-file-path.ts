import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

/**
 * @param relativeFilePath - (e.g. "./image.png")
 * @param absoluteUrlOfTestFile - (e.g. import.meta.url)
 *
 * Example:
 * ```
        await getFilePath("./image.png", import.meta.url);
   ```
 */
export const getFilePath = (relativeFilePath: string, absoluteUrlOfTestFile: string) =>
  resolve(dirname(fileURLToPath(absoluteUrlOfTestFile)), relativeFilePath);
