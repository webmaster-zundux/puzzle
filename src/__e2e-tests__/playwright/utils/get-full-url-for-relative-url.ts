import { type Page } from "@playwright/test";

export const getFullUrlForRelativeUrl = (page: Page, relativeUrl: string) => {
  const baseUrl = page.url();
  const rootPart = baseUrl.substring(0, baseUrl.length - 1);
  const relativePart = relativeUrl.substring(1);

  return `${rootPart}/${relativePart}`;
};
