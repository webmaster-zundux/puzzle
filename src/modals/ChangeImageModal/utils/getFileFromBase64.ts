export const urlToFile = async (url: string, filename: string, mimeType: string) => {
  const res = await fetch(url, {
    mode: "no-cors",
    referrerPolicy: "no-referrer",
    credentials: "omit",
    redirect: "error",
  });
  const buf = await res.arrayBuffer();

  return new File([buf], filename, { type: mimeType });
};
