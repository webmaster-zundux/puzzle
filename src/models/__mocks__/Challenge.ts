export const generateChallengeId = () => {
  const newUUID = "364bf6d0-6686-4892-be7d-618ac2fcdcc7";
  return newUUID;
};
