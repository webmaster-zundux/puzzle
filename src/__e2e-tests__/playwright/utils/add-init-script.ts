import { type Page } from "@playwright/test";
import { getFilePath } from "./get-file-path";

/**
 * @param page - playwright.Page
 * @param relativePathToPreloadScript - (e.g. "./preload-data.js")
 * @param absoluteUrlOfTestFile - (e.g. import.meta.url)
 *
 * Example:
 * ```
        await addInitScript(page, "./init-script.js", import.meta.url);
   ```
 */
export async function addInitScript(page: Page, relativePathToPreloadScript: string, absoluteUrlOfTestFile: string) {
  const filePath = getFilePath(relativePathToPreloadScript, absoluteUrlOfTestFile);

  await page.addInitScript({ path: filePath });
}
