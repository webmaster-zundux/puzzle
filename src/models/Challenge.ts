import type { Id } from "../core/puzzle/BaseEntity";
import { uuid } from "../utils/uuid";

export type ChallengeId = Id;

const generateChallengeIdFunction = () => {
  const newUUID = uuid();
  return newUUID;
};

const EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT =
  import.meta.env.VITE_TESTING_CHALLENGE_ID_GENERATION_EXPOSE_TO_WINDOW_OBJECT === "1";
if (EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT) {
  // @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
  if (typeof window?.generateChallengeId !== "function") {
    Object.defineProperty(window, "generateChallengeId", {
      value: generateChallengeIdFunction,
      writable: false,
      configurable: false,
    });
  }
}

// @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
export const generateChallengeId = window?.generateChallengeId ?? generateChallengeIdFunction;
