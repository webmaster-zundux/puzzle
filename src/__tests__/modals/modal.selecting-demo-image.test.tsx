import { fireEvent, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { App } from "../../App";
import { urlToFile } from "../../modals/ChangeImageModal/utils/getFileFromBase64";
import * as Challenge from "../../models/Challenge";
import { renderIntoDocumentBody } from "../__tests-utils__/renderIntoDocumentBody";
import {
  image250x250Base64,
  isChooseMethodToCreatePuzzleDialogOpened,
  isLoadImageFromLinkDialogOpened,
  isPuzzleCreationDialogOpened,
  isUploadImageFromDeviceDialogOpened,
  openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog,
} from "./modal-helpers";

describe(`modal dialog "choose a demo image to create a puzzle"`, () => {
  describe(`changes it's visibility state`, () => {
    it(`opens when clicking on "demo images" link in "choose method to create puzzle" dialog`, async () => {
      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();

      await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog({ isShortCheck: false });

      expect(renderWithoutDialog).toMatchDiffSnapshot(
        asFragment(),
        {
          aAnnotation: "without dialog",
          bAnnotation: "with dialog",
        },
        `"selecting demo image" dialog`,
      );
    });

    it("closes when clicking on the dialog backdrop", async () => {
      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();
      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog();
      const backdrop = within(dialog).getByLabelText(/modal-backdrop/i);
      expect(backdrop).toBeInTheDocument();

      await userEvent.click(backdrop!);

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without dialog`);
    });

    it('closes when pressing on "esc" key', async () => {
      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();
      await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog();

      await userEvent.keyboard("{Escape}");

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without dialog`);
    });

    it('closes when clicking on "close" button', async () => {
      vi.spyOn(Challenge, "generateChallengeId")
        .mockImplementationOnce(() => "b47663e4-95ab-4472-9344-0f165a6a7ff9")
        .mockImplementationOnce(() => "89e80c0c-186a-46db-b242-b147c88a2a3e");

      const { asFragment } = renderIntoDocumentBody(<App />);
      const renderWithoutDialog = asFragment();
      expect(Challenge.generateChallengeId).toHaveBeenCalledOnce();
      const initialChallengeSeed = screen.getByRole("note", {
        name: /challenge seed:/i,
      }).textContent;
      expect(initialChallengeSeed).toBe("b47663e4-95ab-4472-9344-0f165a6a7ff9");
      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog();

      await userEvent.click(within(dialog).getByRole("button", { name: /close/i }));

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without dialog`);
      expect(Challenge.generateChallengeId).toHaveBeenCalledOnce();
      const challengeSeedAfterCanceledRestartOfPuzzle = screen.getByRole("note", {
        name: /challenge seed:/i,
      }).textContent;
      expect(challengeSeedAfterCanceledRestartOfPuzzle).toBe(initialChallengeSeed);
    });

    it('switches to "choose method to create a puzzle" dialog when clicking on "choose another image" button', async () => {
      renderIntoDocumentBody(<App />);

      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog();
      await userEvent.click(within(dialog).getByRole("button", { name: /choose another image/i }));
      await isChooseMethodToCreatePuzzleDialogOpened();

      expect(screen.queryByRole("dialog", { name: /choose a demo image/i })).not.toBeInTheDocument();
    });

    it(`switches to "load an image from device" dialog when clicking on "load an image from device" link`, async () => {
      renderIntoDocumentBody(<App />);

      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog();
      await userEvent.click(within(dialog).getByRole("link", { name: /load an image from device/i }));
      await isUploadImageFromDeviceDialogOpened();

      expect(screen.queryByRole("dialog", { name: /selecting demo image/i })).not.toBeInTheDocument();
    });

    it(`switches to "load an image from a link" dialog when clicking on "load an image from a link" link`, async () => {
      renderIntoDocumentBody(<App />);

      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog();
      await userEvent.click(within(dialog).getByRole("link", { name: /load an image from a link/i }));
      await isLoadImageFromLinkDialogOpened();

      expect(screen.queryByRole("dialog", { name: /selecting demo image/i })).not.toBeInTheDocument();
    });
  });

  describe.skip(`selecting a demo image`, () => {
    it(`switches to "puzzle creation" dialog when clicking the first demo image button and then clicking "use selected image" button`, async () => {
      const user = userEvent;
      renderIntoDocumentBody(<App />);
      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog({ user });
      const firstDemoImageButton = within(dialog).getByRole("button", {
        name: /\/demo-images\/demo-image-1920x1080\.png/i,
      });
      const imageLink = "/demo-images/demo-image-1920x1080.png";

      await user.click(firstDemoImageButton);
      expect(within(dialog).getByRole("button", { name: /use selected image/i })).toBeEnabled();
      await user.click(within(dialog).getByRole("button", { name: /use selected image/i }));

      await waitFor(() => {
        expect(within(dialog).getByRole("img", { name: /preview of the puzzle cut into pieces/i })).toHaveAttribute(
          "src",
          imageLink,
        );
      });
      await isPuzzleCreationDialogOpened({ isShortCheck: true });
    });

    it(`switches to "puzzle creation" dialog when clicking the second demo image button and then clicking "use selected image" button`, async () => {
      const user = userEvent;
      renderIntoDocumentBody(<App />);
      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog({ user });
      const secondDemoImageButton = within(dialog).getByRole("button", {
        name: /\/demo-images\/demo-image-1080x1920\.png/i,
      });
      const imageLink = "/demo-images/demo-image-1080x1920.png";

      await user.click(secondDemoImageButton);
      expect(within(dialog).getByRole("button", { name: /use selected image/i })).toBeEnabled();
      await user.click(within(dialog).getByRole("button", { name: /use selected image/i }));

      await waitFor(() => {
        expect(within(dialog).getByRole("img", { name: /preview of the puzzle cut into pieces/i })).toHaveAttribute(
          "src",
          imageLink,
        );
      });
      await isPuzzleCreationDialogOpened({ isShortCheck: true });
    });
  });

  describe.skip(`image upload`, () => {
    it(`switches to "puzzle creation" dialog when drag and drop an image file into browser window`, async () => {
      const user = userEvent;
      renderIntoDocumentBody(<App />);
      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog({ user });
      const imageFile = await urlToFile(image250x250Base64, "250x250.png", "image/png");
      const dataTransferItemList = [
        {
          kind: "file",
          type: "image/png",
          getAsFile: () => imageFile,
        },
      ];
      const documentBody = dialog.ownerDocument.body;

      fireEvent.drop(documentBody, { dataTransfer: { items: dataTransferItemList } });

      await waitFor(() => {
        expect(within(dialog).getByRole("img", { name: /preview of the puzzle cut into pieces/i })).toHaveAttribute(
          "src",
          image250x250Base64,
        );
      });
      await isPuzzleCreationDialogOpened({ isShortCheck: true });
    });

    it(`switches to "puzzle creation" dialog when insert an image file from clipboard (hotkey: ctrl + v) into browser window`, async () => {
      const user = userEvent;
      renderIntoDocumentBody(<App />);
      const dialog = await openChooseDemoImageDialogByClickOnRelatedLinkInChooseMethodToCreatePuzzleDialog({ user });
      const imageFile = await urlToFile(image250x250Base64, "250x250.png", "image/png");
      const dataTransferItemList = [
        {
          kind: "file",
          type: "image/png",
          getAsFile: () => imageFile,
        },
      ];
      const documentBody = dialog.ownerDocument.body;

      fireEvent.paste(documentBody, { clipboardData: { items: dataTransferItemList } });

      await waitFor(() => {
        expect(within(dialog).getByRole("img", { name: /preview of the puzzle cut into pieces/i })).toHaveAttribute(
          "src",
          image250x250Base64,
        );
      });
      await isPuzzleCreationDialogOpened({ isShortCheck: true });
    });
  });
});
