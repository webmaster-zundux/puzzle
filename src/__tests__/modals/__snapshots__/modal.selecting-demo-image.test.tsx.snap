// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`src/__tests__/modals/modal.selecting-demo-image.test.tsx > modal dialog "choose a demo image to create a puzzle" > changes it's visibility state > closes when clicking on "close" button: without dialog 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`src/__tests__/modals/modal.selecting-demo-image.test.tsx > modal dialog "choose a demo image to create a puzzle" > changes it's visibility state > closes when clicking on the dialog backdrop: without dialog 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`src/__tests__/modals/modal.selecting-demo-image.test.tsx > modal dialog "choose a demo image to create a puzzle" > changes it's visibility state > closes when pressing on "esc" key: without dialog 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`src/__tests__/modals/modal.selecting-demo-image.test.tsx > modal dialog "choose a demo image to create a puzzle" > changes it's visibility state > opens when clicking on "demo images" link in "choose method to create puzzle" dialog: "selecting demo image" dialog 1`] = `
"Snapshot Diff:
- without dialog
+ with dialog

@@ -453,6 +453,190 @@
            </div>
          </div>
        </aside>
      </main>
    </div>
+   <div>
+     <div
+       aria-label="Choose a demo image"
+       aria-modal="true"
+       class="_Dialog_9b989f _DialogAsDiv_9b989f"
+       role="dialog"
+     >
+       <div
+         class="_DialogBodyContainer_9b989f"
+       >
+         <div
+           aria-hidden="true"
+           aria-label="modal-backdrop"
+           class="_Backdrop_9b989f"
+         />
+         <div
+           class="_DialogBody_9b989f _WithCardForContent_9b989f"
+         >
+           <div
+             class="_CloseIcon_bea9a0"
+           >
+             <button
+               aria-label="close"
+               class="_ActionButton_b70b4c _NoBackground_b70b4c _NoBorder_b70b4c"
+               type="button"
+             >
+               <div
+                 class="_Icon_14488b _DoubleSize_14488b"
+                 data-src="/assets/fa-icons/xmark.svg"
+               >
+                 inline-svg-icon-mock
+               </div>
+             </button>
+           </div>
+           <div
+             class="_DialogContent_9b989f"
+           >
+             <div
+               class="_Content_a024cf"
+             >
+               <h3
+                 class="_Title_a024cf"
+               >
+                 Choose a demo image
+               </h3>
+               <div
+                 class="_MainContent_a024cf"
+               >
+                 <form
+                   class="_Form_bc14b1"
+                 >
+                   <ul
+                     class="_ImageList_bc14b1"
+                   >
+                     <li
+                       class="_ImageListItem_bc14b1"
+                     >
+                       <div
+                         class="_ImageThumbnailContainer_5c0cbc"
+                       >
+                         <button
+                           class="_ActionButton_b70b4c"
+                           style="padding: 0.5em;"
+                           type="button"
+                         >
+                           <img
+                             alt="/demo-images/demo-image-1920x1080.png"
+                             class="_Thumbnail_5c0cbc"
+                             src="/demo-images/demo-image-1920x1080.png"
+                           />
+                         </button>
+                         <div
+                           class="_ImageHumanizedName_5c0cbc"
+                         >
+                           demo image 1920x1080
+                         </div>
+                         <div
+                           class="_ImageSourceLink_5c0cbc"
+                         >
+                           <span>
+                             Image link:
+                           </span>
+                            
+                           <a
+                             href="/demo-images/demo-image-1920x1080.png"
+                             target="__blank"
+                           >
+                             /demo-images/demo-image-1920x1080.png
+                           </a>
+                         </div>
+                       </div>
+                     </li>
+                     <li
+                       class="_ImageListItem_bc14b1"
+                     >
+                       <div
+                         class="_ImageThumbnailContainer_5c0cbc"
+                       >
+                         <button
+                           class="_ActionButton_b70b4c"
+                           style="padding: 0.5em;"
+                           type="button"
+                         >
+                           <img
+                             alt="/demo-images/demo-image-1080x1920.png"
+                             class="_Thumbnail_5c0cbc"
+                             src="/demo-images/demo-image-1080x1920.png"
+                           />
+                         </button>
+                         <div
+                           class="_ImageHumanizedName_5c0cbc"
+                         >
+                           demo image 1080x1920
+                         </div>
+                         <div
+                           class="_ImageSourceLink_5c0cbc"
+                         >
+                           <span>
+                             Image link:
+                           </span>
+                            
+                           <a
+                             href="/demo-images/demo-image-1080x1920.png"
+                             target="__blank"
+                           >
+                             /demo-images/demo-image-1080x1920.png
+                           </a>
+                         </div>
+                       </div>
+                     </li>
+                   </ul>
+                   <div
+                     class="_FormFooter_bc14b1"
+                   >
+                     <div
+                       class="_ActionButtonList_bc14b1"
+                     >
+                       <button
+                         class="_ActionButton_b70b4c _ModalActionButton_b70b4c"
+                         type="button"
+                       >
+                         Choose another image
+                       </button>
+                       <button
+                         class="_ActionButton_b70b4c _PrimaryButton_b70b4c _ModalActionButton_b70b4c"
+                         disabled=""
+                         type="submit"
+                       >
+                         Use selected image
+                       </button>
+                     </div>
+                     <div
+                       aria-label="alternative way to load an image"
+                       class="_Note_51962b"
+                       role="note"
+                     >
+                       also you can 
+                       <a
+                         class="_NoteLink_edcc23"
+                         href="#"
+                       >
+                         upload an image from device
+                       </a>
+                        
+                       <br />
+                       or 
+                       <a
+                         class="_NoteLink_edcc23"
+                         href="#"
+                       >
+                         load an image from a link
+                       </a>
+                        
+                       <br />
+                     </div>
+                   </div>
+                 </form>
+               </div>
+             </div>
+           </div>
+         </div>
+       </div>
+     </div>
+   </div>
  </DocumentFragment>"
`;
